### PR TITLE
Support for NEURON {RANDOM ranvar1, ranvar2, ...} statement.

### PR DESCRIPTION
**Pull request by nrnhines**
Accepts the latest netcon.mod from neuronsimulator/nrn#2627.  Some changes were picked from #581

- [x] Misuse of RANDOM variables should result in meaningful error messages
- [x] Unit tests
- [x] Semantic analysis visitor improvements

### CODE REVIEW

**Comment by nrnhines** [comment]
Note that, at the moment, the manual change to netstim.cpp is (still incomplete but compiles).
```
hines@hines-t7500:~/neuron/rannmodl$ diff buildsani/bin/x86_64/corenrn/mod2c/netstim.cpp netstim.cpp
7c7
< Created         : Thu Dec 14 19:14:16 2023
---
> Created         : Thu Dec 14 13:52:59 2023
29d28
< 
46a46
>         "ranvar",
49a50
> #define Zranvar (nrnran123_State*)inst->ranvar[indexes[2*pnodecount + id]]
77a79
>         void** ranvar{};
82d83
< 
96c97
<         return -1;
---
>         return 2;
111c112
<         return 3;
---
>         return 4;
191a193
>         inst->ranvar = nt->_vdata;
267c269
<                         erand_in_0 = random_negexp(ranvar);
---
>                         erand_in_0 = nrnran123_negexp(Zranvar);
283c285
<         random_setseq(ranvar, x);
---
>         nrnran123_setseq(Zranvar, x);
298c300
<                 nrnran123_setids(ranvar, id[0], id[1], id[2]);
---
>                 nrnran123_setids(Zranvar, id[0], id[1], id[2]);
301c303
<                 nrnran123_setseq(ranvar, id[0], which);
---
>                 nrnran123_setseq(Zranvar, id[0], which);
314c316
<                 nrnran123_setids(ranvar, static_cast<uint32_t>(*getarg(1)), static_cast<uint32_t>(*getarg(2)), static_cast<uint32_t>(*getarg(3)));
---
>                 nrnran123_setids(Zranvar, static_cast<uint32_t>(*getarg(1)), static_cast<uint32_t>(*getarg(2)), static_cast<uint32_t>(*getarg(3)));
316c318
<                 nrnran123_setids(ranvar, static_cast<uint32_t>(*getarg(1)), static_cast<uint32_t>(*getarg(2)), 0);
---
>                 nrnran123_setids(Zranvar, static_cast<uint32_t>(*getarg(1)), static_cast<uint32_t>(*getarg(2)), 0);
318c320
<             nrnran123_setseq(ranvar, 0, 0);
---
>             nrnran123_setseq(Zranvar, 0, 0);
335c337
<                 erand_in_0 = random_negexp(ranvar);
---
>                 erand_in_0 = nrnran123_negexp(Zranvar);
345c347
<         ret_erand = random_negexp(ranvar);
---
>         ret_erand = nrnran123_negexp(Zranvar);
393c395
<                                         erand_in_0 = random_negexp(ranvar);
---
>                                         erand_in_0 = nrnran123_negexp(Zranvar);
441c443
<                                     erand_in_0 = random_negexp(ranvar);
---
>                                     erand_in_0 = nrnran123_negexp(Zranvar);
482c484
<                     random_setseq(ranvar, x_in_0);
---
>                     nrnran123_setseq(Zranvar, x_in_0);
506c508
<                                 erand_in_0 = random_negexp(ranvar);
---
>                                 erand_in_0 = nrnran123_negexp(Zranvar);
537c539,540
<         hoc_register_dparam_semantics(mech_type, 2, "netsend");
---
>         hoc_register_dparam_semantics(mech_type, 2, "ranvar");
>         hoc_register_dparam_semantics(mech_type, 3, "netsend");
hines@hines-t7500:~/neuron/rannmodl$ 
```

File: src/codegen/codegen_coreneuron_cpp_visitor.cpp [hunk 1]
```diff
@@ -2364,7 +2388,11 @@ void CodegenCoreneuronCppVisitor::print_instance_variable_setup() {
             if (var.is_index || var.is_integer) {
                 return "ml->pdata";
             } else if (var.is_vdata) {
-                return "nt->_vdata";
+                if (var.symbol->has_all_properties(NmodlType::random_var)) {
+                    return "(nrnran123_State**)nt->_vdata";
```
Comment (nrnhines) [inline src/codegen/codegen_coreneuron_cpp_visitor.cpp:2392]: Is this how we "cast" these days. Or should it be more decorated?

File: src/visitors/semantic_analysis_visitor.cpp [hunk 2]
```diff
@@ -6,8 +6,10 @@
  */
 
 #include "visitors/semantic_analysis_visitor.hpp"
+#include "ast/statement_block.hpp"
```
Comment (nrnhines) [inline src/visitors/semantic_analysis_visitor.cpp:9]: I wonder if we should just get all of them.

File: src/visitors/semantic_analysis_visitor.cpp [hunk 3]
```diff
@@ -41,6 +44,17 @@ bool SemanticAnalysisVisitor::check(const ast::Program& node) {
     const auto& sym_table = node.get_symbol_table();
     assert(sym_table != nullptr);
 
+    // this block is going nowhere since nodes below is just its declaration
```
Comment (nrnhines) [inline src/visitors/semantic_analysis_visitor.cpp:47]: This whole block should be removed. I left it just to remind me to mention a confusion I had at the beginning of my implementation where I thought ```get_nodes()``` would return every node where a particular variable appeared.

File: src/visitors/semantic_analysis_visitor.cpp [hunk 4]
```diff
@@ -89,6 +103,99 @@ void SemanticAnalysisVisitor::visit_function_block(const ast::FunctionBlock& nod
...
+            auto parent = node.get_parent();
+            if (parent) {
+                if (parent->is_var_name()) {
+                    parent = parent->get_parent();
+                    if (parent && parent->is_function_call()) {
+                        // The function must be a random function
+                        auto fname = parent->get_node_name();
+                        if (symtab::syminfo::is_random_function(fname)) {
+                            // but is name the first arg?
+                            ast::FunctionCall* rfun = (ast::FunctionCall*) parent;
```
Comment (nrnhines) [inline src/visitors/semantic_analysis_visitor.cpp:134]: Should this cast be much more elaborate?

**Review by pramodk — COMMENTED** [review]
@nrnhines: tomorrow I am going to work a bit on `src/visitors/semantic_analysis_visitor.*` and address `Todo` about `meaningful error` message.

File: src/visitors/semantic_analysis_visitor.cpp (hunk 3 above)
Comment (pramodk) [reply src/visitors/semantic_analysis_visitor.cpp]: Ok, removed.

File: src/visitors/semantic_analysis_visitor.cpp (hunk 4 above)
Comment (pramodk) [reply src/visitors/semantic_analysis_visitor.cpp]: C++ `dynamic_cast` is sufficient here.

File: src/codegen/codegen_coreneuron_cpp_visitor.cpp (hunk 1 above)
Comment (pramodk) [reply src/codegen/codegen_coreneuron_cpp_visitor.cpp]: I implemented the above suggestion.

**Review by pramodk — APPROVED** [review]
LGTM - I have made the necessary changes directly.

**Pull request merged**